thorlcr/master/thgraphmanager.cpp

```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
+
+        virtual void processDebugCommand(CSafeSocket &ssock, StringBuffer &rawText)
+        {
+            Owned<IPropertyTree> queryXml;
+            try
+            {
+                queryXml.setown(createPTreeFromXMLString(rawText.str(), ipt_caseInsensitive, (PTreeReaderOptions)(ptr_ignoreWhiteSpace|ptr_ignoreNameSpaces)));
+            }
+            catch (IException *E)
+            {
+                StringBuffer s;
+                DBGLOG("ERROR: Invalid XML received from %s:%s", E->errorMessage(s).str(), rawText.str());
+                throw;
+            }
+
+            Linked<CJobMaster> job = mgr.getCurrentJob();
+            if (!job)
+                throw MakeStringException(5300, "Command not available when no graph active");//TODO:Update error code
+
+            const char *command = queryXml->queryName();
+            const char *edgeId = queryXml->queryProp("@edgeId");
+            const char *graphId = job->queryGraphName()?job->queryGraphName():"";
+
+            if (!command || !edgeId) throw MakeStringException(5300, "Invalid debug command");
+
+            enum msgids cmdcode;
+            if (strncmp(command,"print", 5) == 0)
+                cmdcode = DebugRequest;
+            else if (strncmp(command,"quit", 4) == 0)
+                cmdcode = GraphAbort;
+            else
+                throw MakeStringException(5300, "Command not supported by Thor");
+
+            FlushingStringBuffer response(&ssock, false, MarkupFmt_XML, false, false, queryDummyContextLogger());
+            response.startDataset("Debug", NULL, (unsigned) -1);
+
+            ICommunicator &comm = job->queryNodeComm();
+            CMessageBuffer mbuf;
+            mbuf.append(cmdcode);
+            mbuf.append(job->queryKey());
+            mptag_t replyTag = createReplyTag();
+            serializeMPtag(mbuf, replyTag);
+            mbuf.append(rawText);
+            if (!comm.send(mbuf, RANK_ALL_OTHER, masterSlaveMpTag, MP_ASYNC_SEND))
+            {
+                DBGLOG("Failed to send debug info to slave");
+                throwUnexpected();
+            }
+            unsigned nodes = job->queryNodes();
+            response.appendf("<%s graphId='%s' edgeId='%s'>", command, graphId, edgeId);
+            while (nodes)
+            {
+                rank_t sender;
+                mbuf.clear();
+                comm.recv(mbuf, RANK_ALL, replyTag, &sender, 10000);
+                while (mbuf.remaining())
+                {
+                    StringAttr row;
+                    mbuf.read(row);
+                    response.append(row);
+                }
+                nodes--;
+            }
+            response.append("</print>");
+            response.flush(true);
+        }
+        virtual void main()
+        {
+            sock.setown(ISocket::create(port));
+            while (running)
+            {
+                try
+                {
+                    ISocket *client = sock->accept(true);
```

Comment:
picky (and debatable): Think looks clearer if this is a Owned<ISocket> and then the CSafeSocket line below is: CSafeSocket ssock(client.getClear());